core/src/test/scala/unit/kafka/security/auth/ZkAuthorizationTest.scala

```diff
@@ -0,0 +1,265 @@
+  }
+
+  @Test
+  def testZkAntiMigration() {
+    val unsecureZkUtils = ZkUtils(zkConnect, 6000, 6000, false)
+    try {
+      testMigration(zkUtils, unsecureZkUtils)
+    } finally {
+      unsecureZkUtils.close()
+    }
+  }
+  
+  @Test
+  def testDelete() {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- zkUtils.securePersistentZkPaths) {
+      info("Creating " + path)
+      zkUtils.makeSurePersistentPathExists(path)
+    }
+    zkUtils.zkConnection.setAcl("/", zkUtils.DefaultAcls, -1)
+    System.setProperty(JaasUtils.ZK_SASL_CLIENT, "false")
+    val unsecureZkUtils = ZkUtils(zkConnect, 6000, 6000, false)
+    var failed = false
+    for (path <- unsecureZkUtils.securePersistentZkPaths) {
+      info("Deleting " + path)
+      try {
+        unsecureZkUtils.deletePath(path)
+        info("Failed for path %s".format(path))
+        failed = true
+      } catch {
+        case e: Exception => // Expected
+      }
+    }
+    unsecureZkUtils.close()
+    System.clearProperty(JaasUtils.ZK_SASL_CLIENT)
+    if (failed)
+      fail("Was able to delete at least one znode")
+  }
+
+  @Test
+  def testDeleteRecursive() {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- zkUtils.securePersistentZkPaths) {
+      info("Creating " + path)
+      zkUtils.makeSurePersistentPathExists(path)
+      zkUtils.createPersistentPath(path + "/fpjwashere", "")
+    }
+    zkUtils.zkConnection.setAcl("/", zkUtils.DefaultAcls, -1)
+    System.setProperty(JaasUtils.ZK_SASL_CLIENT, "false")
+    val unsecureZkUtils = ZkUtils(zkConnect, 6000, 6000, false)
+    for (path <- unsecureZkUtils.securePersistentZkPaths) {
+      info("Deleting " + path)
+      try {
+        unsecureZkUtils.deletePath(path)
+        fail("Should have thrown an exception: %s".format(path))
+      } catch {
+        case e: Exception => // Expected
+      }
+      try {
+        unsecureZkUtils.deletePath(path + "/fpjwashere")
+        fail("Should have thrown an exception: %s".format(path))
+      } catch {
+        case e: Exception => // Expected
+      }
+    }
+    unsecureZkUtils.close()
+    System.clearProperty(JaasUtils.ZK_SASL_CLIENT)
+  }
+
+  private def testMigration(firstZk: ZkUtils, secondZk: ZkUtils) {
+    info("zkConnect string: %s".format(zkConnect))
+    for (path <- firstZk.securePersistentZkPaths) {
+      info("Creating " + path)
+      firstZk.makeSurePersistentPathExists(path)
+      // Create a child for each znode to exercise the recurrent
+      // traversal of the data tree
+      firstZk.createPersistentPath(path + "/fpjwashere", "")
+    }
+    val secureOpt: String  = secondZk.isSecure match {
+      case true => "secure"
+      case false => "unsecure"
+    }
+    ZkSecurityMigrator.run(Array("--zookeeper.acl=%s".format(secureOpt), "--zookeeper.connect=%s".format(zkConnect)))
+    info("Done with migration")
+    for (path <- secondZk.securePersistentZkPaths) {
+      secondZk.makeSurePersistentPathExists(path)
```

Comment:
Not strictly necessary, it is exercising the code path when the znode already exists. I removed it.